feat: linting the start of declarations and commands

adomani opened the pull request:
This linter checks that
* each command starts at the beginning of a line;
* the "hypotheses segment" of each declaration coincides with its pretty-printed version.

This is likely to have still false positives, but maybe it is acceptable!

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

adomani committed:
feat: linting the start of declarations and commands

adomani committed:
lnie breaks in tests

adomani committed:
chore(MathlibTest/order): reduce imports and formatting

adomani committed:
add tests

adomani committed:
docs

grunweg commented on MathlibTest/PPRoundtrip.lean:
```diff
@@ -64,3 +65,21 @@ set_option linter.ppRoundtrip true in
+note: this linter can be disabled with `set_option linter.style.commandStart false`
+-/
+#guard_msgs in
+example {a :Nat} : a = a := rfl
```
Does the linter also catch the following cases/can you add those as well, please?
```
example  {a :Nat} : a = a := rfl
example {a : Nat}{b : Nat} : a = a := rfl
example {a : Nat}  : a = a := rfl
example {alpha: Type*} [Neg alpha   ] {a : Nat} : a = a := rfl

grunweg reviewed (commented):
Thanks for starting this discussion! I wonder about a few more cases - can you add them as tests if they are already covered, please?

adomani committed:
add tests

adomani committed:
move tests to correct file

adomani committed:
docs

adomani committed:
lint also variable commands

adomani committed:
actually lint

grunweg commented:
Eventually, you want to import this linter in `Mathlib.Init`, right? (The current state is a decent start, it looks like there will be enough things to fix already.) How hard is it to write an auto-fixing script for the linter violations, to help with that?

adomani committed:
some comment removal

adomani committed:
chore: more whitespace adaptation

adomani committed:
chore: more whitespace adaptation

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/pplint_some

adomani committed:
Merge branch 'whitespace/more_adaptations' into adomani/pplint_some

adomani committed:
bonus

adomani committed:
Merge branch 'whitespace/more_adaptations' into adomani/pplint_some

adomani committed:
chore: spaces in LinearCombination'

adomani committed:
do not lint subtype notation

adomani committed:
unlintedNodes

adomani committed:
more

adomani committed:
chore: more whitespace fixes

adomani committed:
chore: more whitespace fixes

adomani committed:
Merge branch 'whitespace/more_fixed_files' into adomani/pplint_some

adomani committed:
Merge branch 'whitespace/linear_combination_prime' into adomani/pplint_some

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/pplint_some

adomani committed:
verbose and add :

adomani committed:
Merge branch 'adomani/pplint_some' of github.com:leanprover-community/mathlib4 into adomani/pplint_some

adomani committed:
fix guard_msgs in probability

adomani committed:
chore(whitespace): lots of whitespace

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/pplint_some

adomani committed:
chore(whitespace): some more changes in whitespace

adomani committed:
fix: doc-string vs comment and guillemets

adomani committed:
fix test

adomani committed:
list notation

adomani committed:
drop entirely `/-` comment lines, even if they end before the end of the line or continue

adomani committed:
fixme for doc-string truncation

adomani committed:
fix guard_msgs

adomani committed:
chore(whitespace): more whitespace changes

adomani committed:
Merge branch 'master' into adomani/pplint_some

adomani committed:
two more

adomani committed:
chore(whitespace): more changes

adomani committed:
chore: remove porting note

adomani committed:
chore(whitespace): yet some more whitespace changes

adomani committed:
chore(whitespace): changing yet more spaces

adomani committed:
unlint fiber bundles and slice

adomani committed:
chore: adapt more whitespace

adomani committed:
on parsing errors, return

adomani committed:
fix error message

adomani committed:
Use correct names

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/pplint_some

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/pplint_some

adomani committed:
chore: whitespace again

adomani committed:
chore: whitespace again

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/pplint_some

adomani committed:
Merge branch 'whitespace/here_we_go_again5' into adomani/pplint_some

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/pplint_some

adomani committed:
docs and backticks

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/pplint_some

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/pplint_some

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/pplint_some

grunweg committed:
Merge branch 'master' into adomani/pplint_some

grunweg committed:
chore: annotate four false positives

grunweg committed:
chore: fix some double spaces

grunweg committed:
chore: miscellaneous style fixes

grunweg committed:
chore: make some comments doc comments

Noticed by the linter in #22760: these seem useful as doc comments.

grunweg committed:
one more false positive, todo fold

grunweg committed:
chore(QPF/Multivariate/Basic): remove stray comment

Pointed out by the linter in #22670; my impression is that the comment
can simply be removed.

grunweg committed:
chore: annotate false positives due to C^n notation

I am not sure why this does not error on the R^n notation; perhaps a pretty-printer
ought to be fixed instead.

grunweg commented:
With the above fixes, I believe things are fine: let's see if I overlooked something when scrolling.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

grunweg committed:
Merge branch 'master' into adomani/pplint_some

grunweg commented on Mathlib/Tactic/Linter/CommandStart.lean:
```diff
@@ -0,0 +1,149 @@
+      if let some pos := stype.getPos? then
+        if pos.1 ≤ finalLintPos.1 then
+          return
+    let stx := capSyntax stx finalLintPos.1
```
I didn't scrutinise the tests file yet: it would be great to test this line, to clarify the behaviour. (Basically, my ideal is that if the tests pass, a linter change is correct: if mathlib exposes further edge cases, adding them as tests would be great.) With that in mind, does the file already have tests for
- unlinted nodes being left alone
- syntax before an unlinted node gets linted?

grunweg commented on Mathlib/Topology/Algebra/UniformField.lean:
```diff
@@ -94,6 +94,7 @@ open Classical in
 The value of `hat_inv` at zero is not really specified, although it's probably zero.
 Here we explicitly enforce the `inv_zero` axiom.
 -/
+open Classical in
```
This was a bad merge on my part:
```suggestion

grunweg commented on Mathlib/Tactic/Linter/CommandStart.lean:
```diff
@@ -0,0 +1,149 @@
+        let diff := real.firstDiffPos st
+        let pos := posToShiftedPos lths diff.1 + origSubstring.startPos.1
+        let f := origSubstring.str.drop (pos)
+        let extraLth := (f.takeWhile (· != st.get diff)).length
```
This is my only major comment - and it's somewhat pre-existing from the ppRoundtrip linter (but relevant now, that I'd like this linter to be enabled by default). The simplest thing to do would be to show the old and new lines; I appreciate that the linter is not doing just that (as "eyeball this line to find the change" works surprisingly well, but can be tedious). 

The zooming in around the diff shows a short snippet of what changes (which makes the change easier to spot) --- but the strict length of just seven characters (in the simplest case) gives suggestions which are also hard to parse. `Current syntax:  'le {a: Nat'`, `Expected syntax: 'le {a : Na'` is simply not pretty.
Do you see a reasonable way to at least print "full units" (e.g., up to a corresponding closing brackets, for instance)? That's probably hard in general. Typing this now, though, perhaps a heuristic of "extend until up the corresponding opening/closing bracket" could already work well.

What do you think?

grunweg reviewed (commented):
I took a first detailed look at the implementation. Here's my usual dozen of comments. Would you like some help with e.g. adding some of the tests I asked for?

grunweg commented:
(I just saw that the linter is still work in progress - so you may be aware of all of these. In that case, feel free to take this as an indication "I also find that important".)

grunweg committed:
Merge branch 'master' into adomani/pplint_some

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

adomani commented:
Michael, thank you for the review!  This linter is maturing into #24465.  That linter is much more flexible and almost ready for review.  I just need to audit its exceptions and tweak it a little.

Feel free to take a look at it yourself, if you're willing to PR some of its suggestions or figure out what goes wrong.  Right now, I'm aware of poor handling of comments by the linter.

grunweg commented:
That linter looks great! I'm taking a look at its current errors now, starting from the back of the output (and will push some fixes and some things I noticed there).

adomani commented:
Thank you for all the comments!

For the record (and as you know), I prefer the linter at #24465, so I am closing this PR!
